### PR TITLE
feat: mark explicit content, support multiple artists on Tidal

### PR DESCRIPTION
**Pull request by Im-Beast**
Marks tracks and albums with explicit lyrics as "Explicit".
On top of that I've extracted building articles to their own functions, so it is easier to read and maintain.
I have also ran Black formatter, as I couldn't find any mention of what should be used to format code.

code isn't tested, please actually review :^)

### CODE REVIEW

File: deezer.py [hunk 1]
```diff
@@ -16,48 +16,91 @@ def env_constructor(loader, node):
 
 yaml.SafeLoader.add_constructor("!env", env_constructor)
 
-if __name__ == '__main__':
-    with open("config.yml", 'r') as f:
+
+async def build_track_entry(builder, el):
```
Comment (JuniorJPDJ) [inline deezer.py:20]: probably should be called `add_*_entry` instead of `build_*` as it adds it beyond of just building

Comment (JuniorJPDJ) [reply deezer.py:20]: nvm

Comment (JuniorJPDJ) [reply deezer.py:20]: but wait.. it should at least return it back to the main func

Comment (JuniorJPDJ) [reply deezer.py:20]: but wait.. it should at least return those to the main funcs

Comment (Im-Beast) [reply deezer.py:20]: > but wait.. it should at least return those to the main funcs

fixed

**Comment by JuniorJPDJ** [comment]
tbh I'm not sure if i like extracting those buildings to the separate functions as now almost only thing the track/artist/whatever funcs are doing is http-getting those, do we really need more functions? or just proper formating XD

**Comment by Im-Beast** [comment]
The problem I had is that you can't just put a block instead of expression into `[ expr for y in z ]`, so it would had to be reworked into a normal loop anyways.

I think having them as a separate functions is just less prone to user errors, and also allows for doing more stuff before actually building the article, making things more readable than just inlining everything into that method

**Comment by Im-Beast** [comment]
Also, I don't see track/album only having http requests as an issue, it splits responsibility between them quite nicely, if it is really a big issue I don't mind changing it later :p

**Comment by JuniorJPDJ** [comment]
Hmm... makes sense, lets leave it as it is done by yourself.
I'll checkout it locally to see if it works and I'll let you know

File: tidal.py [hunk 2]
```diff
@@ -17,8 +17,75 @@ def env_constructor(loader, node):
...
+    )
+
+    artists = ", ".join(map(lambda artist: artist["name"], el["artists"]))
+    artists_label = f"Artists: {artists}" if len(artists) > 1 else f"Artist: {artists}"
+
+    album_label = f"Album: {el['album'['title']]}"
+
+    explicit = (
+        el["properties"]["content"] == "explicit"
+        if hasattr(el, "properties") and hasattr(el["properties"], "content")
```
Comment (JuniorJPDJ) [inline tidal.py:35]: for sure not hasattr, you can check for `properties in el` but it's not needed, you can use `explicit = el.get("properties", {}).get("content", "") == "explicit"`

File: tidal.py [hunk 3]
```diff
@@ -17,8 +17,75 @@ def env_constructor(loader, node):
...
+    title = (
+        el["title"]
+        if "version" not in el or not el["version"]
+        else f"{el['title']} ({el['version']})"
+    )
+
+    artists = ", ".join(map(lambda artist: artist["name"], el["artists"]))
+    artists_label = f"Artists: {artists}" if len(artists) > 1 else f"Artist: {artists}"
+
+    album_label = f"Album: {el['album'['title']]}"
```
[the remark this answers is not in the document]
Comment (Im-Beast) [reply tidal.py:31]: 1. I like spacing things out when they are grouped by functionality or something just takes a little bit more space, basically preference. I can make it tighter if you want

2. Oops, will fix tomorrow, I already went to bed :x

If there's anything more let me know, will fix them tomorrow (unless you want to do it yourself, that's fine to me as well)

**Comment by JuniorJPDJ** [comment]
ok, it for sure is bugged, so to make it easy to review please split it into commits providing separate things
1. commit adding your base changes
2. commit extracting builder functions
3. commit for reformat
it's hard to review if not sure which parts of code are changed and which are just copied

File: tidal.py (hunk 2 above)
Comment (Im-Beast) [reply tidal.py:35]: Will do, I was looking for something like this, I'm not really that into python 😁

File: deezer.py [hunk 4]
```diff
@@ -16,48 +16,91 @@ def env_constructor(loader, node):
...
+    artist_label = f"Artist: {el['artist']['name']}"
+    album_label = f"Album: {el['album']['title']}"
+    explicit_label = "\nExplicit" if el["explicit_lyrics"] else ""
```
Comment (JuniorJPDJ) [inline deezer.py:23]: we need to check, but I'm not sure if telegram allows 3 lines of description

File: tidal.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JuniorJPDJ) [reply tidal.py:31]: 1st was just about this empty line above album_label where artists label is not spaced from artists
i'd remove this empty line and if you want you can add empty line after `artists` - labels are one funtionality :D

I'd go with something like:
```
title = ...
artists = ...

artists_label = ...
album_label = ...
```

File: deezer.py (hunk 4 above)
Comment (Im-Beast) [reply deezer.py:23]: Telegram API does not specify max length or amount of lines in an article description, so we have to check that by running the actual code

Comment (JuniorJPDJ) [reply deezer.py:23]: this is exacty what I meant :D

Comment (JuniorJPDJ) [reply deezer.py:23]: just checked and it doesn't show up

**Comment by JuniorJPDJ** [comment]
ok, after testing it together - lets merge
commits are named nicely so lets just rebase and merge

**Pull request merged**